ant/ant.go

```diff
@@ -68,6 +88,39 @@ func clearPorts(config AntConfig) error {
 	return nil
 }
 
+// upgraderThread is a goroutine that waits `UpgradeDelay` and upgrades to the
+// next version in the ant's `UpgradePath`. Once the final version is reached,
+// upgrading is stopped.
+func (a *Ant) upgraderThread() {
+	err := a.tg.Add()
+	if err != nil {
+		return
+	}
+	defer a.tg.Done()
+
+	for _, version := range a.config.UpgradePath {
+		log.Printf("upgrading %v to %v...\n", a.config.Name, version)
+
+		stopSiad(a.APIAddr, a.siad.Process)
+
+		newSiadPath := path.Join(a.config.UpgradeDir, fmt.Sprintf("%v-%v-%v", version, runtime.GOOS, runtime.GOARCH), "siad")
+
+		siad, err := newSiad(newSiadPath, a.config.SiaDirectory, a.config.APIAddr, a.config.RPCAddr, a.config.HostAddr)
+		if err != nil {
+			log.Printf("error starting siad after upgrade: %v\n", err)
+			continue
+		}
+
+		a.siad = siad
```

Comment:
oh, no it actually doesn't since the only place where `a.siad` is read is in `Ant`'s Close() method, which stops the Ant's threadgroup, ending this thread before `a.siad` is read by Close's thread